### PR TITLE
Switch PluralRules Data to ZeroVec

### PR DESCRIPTION
**Pull request by zbraniecki**
^^^^^^^^^^^

### CODE REVIEW

File: components/plurals/src/rules/runtime/ast.rs [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RangeOrValue {
+    Range(u32, u32), // XXX: Can we get away from smaller?
+    Value(u32),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Relation<'data> {
+    pub(crate) and_or: AndOr,
+    pub(crate) polarity: Polarity,
```
Comment (Manishearth) [inline components/plurals/src/rules/runtime/ast.rs:54]: thought (also a note to self, really): if these are refactored into an AndOrPolarityOperand struct on the non-ULE side and we just implement ULE on that (instead of doing the `andor_polarity_operand` field) this becomes something much more amenable to the custom derive that I hope to have

**Review by Manishearth — COMMENTED** [review]
Looks like you're on the right track so far!

**Comment by zbraniecki** [comment]
I applied Manish feedback and tested using exponent for modulo and cutting `RangeOrValueULE` to be `PlainOldULE<4>` - neither really impacts performance (the former deteriorates perf, the latter gives ~1-2% win), so I'd prefer not to overoptimize it right now.

@sffc @Manishearth - can I get second round of feedback? After that, I'd like to polish the code and submit for full review.

File: components/plurals/src/lib.rs [hunk 2]
```diff
@@ -211,26 +211,27 @@ impl PluralCategory {
...
+    /// assert_eq!(categories.next(), Some(PluralCategory::Two));
+    /// assert_eq!(categories.next(), Some(PluralCategory::Few));
+    /// assert_eq!(categories.next(), Some(PluralCategory::Many));
+    /// assert_eq!(categories.next(), Some(PluralCategory::Other));
     /// assert_eq!(categories.next(), None);
     /// ```
     ///
     /// [`Plural Categories`]: PluralCategory
-    pub fn all() -> impl ExactSizeIterator<Item = &'static Self> {
+    pub fn all() -> impl ExactSizeIterator<Item = Self> {
```
Comment (sffc) [inline components/plurals/src/lib.rs:224]: Thought: I would rather have the order change be in its own PR, but I'm not going to stop it from landing here.

File: components/plurals/src/rules/mod.rs [hunk 3]
```diff
@@ -143,13 +144,5 @@
...
-pub(crate) mod lexer;
-pub(crate) mod parser;
-pub(crate) mod resolver;
-pub(crate) mod serializer;
-
-pub use lexer::Lexer;
-pub use parser::{parse, parse_condition};
-pub use resolver::test_condition;
-pub use serializer::serialize;
+pub mod reference;
```
Comment (sffc) [inline components/plurals/src/rules/mod.rs:147]: Question: Why are you making these two modules `pub`? I want to understand your mental model around the visibility of the types in this crate.

File: components/plurals/src/rules/runtime/ast.rs [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+
+impl Operand {
+    fn idx(self) -> u8 {
+        self as u8
+    }
+
+    fn from_idx(idx: u8) -> Self {
+        debug_assert!(idx < 8);
+        // safe due to the repr(u8)
+        unsafe { mem::transmute(idx) }
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:70]: Suggestion (optional): using `num_enum` would remove the `unsafe` from our code.  Also I'm not a big fan of having the `8` statically in our function with no way to tie it to the enum if we add new entries to the enum in the future.

**Review by sffc — COMMENTED** [review]
A few comments to start; more coming soon

File: components/plurals/src/rules/mod.rs (hunk 3 above)
Comment (zbraniecki) [reply components/plurals/src/rules/mod.rs:147]: I imagine people using this crate in one of two ways:

```rust
use icu::plurals::PluralRules;

let pr = PluralRules::try_new(...).expect(...);
pr.select(5);
```

or:

```rust
use icu::plurals::rules::reference::{parse, ast, serialize};

let ast = parse("i % 10 = 5..3").expect(...);
if inspect_ast_to_learn_something(&ast) {
  let new_ast = modify_ast(&ast);
  let string = serialize(ast).expect(...);
  write_to_file(string);
}
```

For the latter, I imagine we may want to put it behind a feature, but since it doesn't loop in any new dependencies, we can just let DCE cut it when not needed.

Does it sound acceptable?

File: components/plurals/src/rules/runtime/ast.rs [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+        // it's hard constructing custom DSTs, we fake a pointer/length construction
+        // eventually we can use the Pointer::Metadata APIs when they stabilize
+        let fake_slice = core::ptr::slice_from_raw_parts(ptr as *const RangeOrValueULE, len_new);
+        let ret = &*(fake_slice as *const Self);
+        debug_assert_eq!(core::mem::size_of_val(ret), core::mem::size_of_val(bytes));
+        ret
+    }
+
+    #[inline]
+    fn as_byte_slice(&self) -> &[u8] {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:372]: Suggestion: I think you can leave `as_byte_slice` auto-implemented ?

File: components/plurals/src/rules/runtime/ast.rs [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+        let operand = Operand::from_idx(encoded & 0b0011_1111);
+        (and_or, polarity, operand)
+    }
+}
+
+unsafe impl VarULE for RelationULE {
+    type Error = &'static str;
+
+    #[inline]
+    unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:358]: Request: please verify this impl with @Manishearth

File: components/plurals/src/rules/runtime/ast.rs [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+                self as *const Self as *const u8,
+                core::mem::size_of_val(self),
+            )
+        }
+    }
+
+    //     what it should do is attempt to parse the first 4 bytes as a u32::ULE (POU<4>), and the remaining bytes as a ZV<RangeOrValueULE>
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let bits = bytes[0]; // XXX: Validate those bits
+        <PlainOldULE<4> as ULE>::validate_byte_slice(&bytes[1..5]).map_err(|_| "foo")?;
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:384]: Nit: `<PlainOldULE<4> as ULE>::validate_byte_slice` is infallible. You can either mark the error case as `unreachable!()` or just delete this line.

File: components/plurals/src/rules/runtime/ast.rs [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+            core::slice::from_raw_parts(
+                self as *const Self as *const u8,
+                core::mem::size_of_val(self),
+            )
+        }
+    }
+
+    //     what it should do is attempt to parse the first 4 bytes as a u32::ULE (POU<4>), and the remaining bytes as a ZV<RangeOrValueULE>
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let bits = bytes[0]; // XXX: Validate those bits
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:383]: Nit: this is an important "XXX" :)

File: components/plurals/src/rules/runtime/ast.rs [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+            and_or,
+            polarity,
+            operand,
+            modulo: u32::from_unaligned(self.modulo),
+            range_list: ZeroVec::borrowed_from_slice(&self.range_list),
+        }
+    }
+
+    fn encode_andor_polarity_operand(and_or: AndOr, polarity: Polarity, operand: Operand) -> u8 {
+        // XXX: Ensure and_or is one bit, polarity is one bit, and operand is max 6 bits
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:328]: and_or and polarity are guaranteed 1 bit because they are booleans.  It can't hurt to add a debug_assert for `operand.idx()`.

File: components/plurals/src/rules/runtime/ast.rs [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+            AndOr::Or
+        };
+
+        let polarity = if encoded & 0b0100_0000 != 0 {
+            Polarity::Positive
+        } else {
+            Polarity::Negative
+        };
+
+        let operand = Operand::from_idx(encoded & 0b0011_1111);
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:349]: See my updated comment above regarding `Operand::from_idx`.  It is correct here to use the infallible (panicking or unsafe) version of that function, since we are inside an `unsafe` function.

File: components/plurals/src/rules/runtime/ast.rs [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+        (and_or, polarity, operand)
+    }
+}
+
+unsafe impl VarULE for RelationULE {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:354]: Nit: Please add a comment with the safety checklist for `VarULE`

File: components/plurals/src/rules/runtime/ast.rs [hunk 12]
```diff
@@ -0,0 +1,634 @@
...
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let bits = bytes[0]; // XXX: Validate those bits
+        <PlainOldULE<4> as ULE>::validate_byte_slice(&bytes[1..5]).map_err(|_| "foo")?;
+        let remaining = &bytes[5..];
+        RangeOrValueULE::validate_byte_slice(remaining).map_err(|_| "foo")?;
+        Ok(())
+    }
+}
+
+unsafe impl EncodeAsVarULE<RelationULE> for Relation<'_> {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:391]: Nit: Please add a safety comment for `unsafe impl`s like this one

File: components/plurals/src/rules/runtime/ast.rs [hunk 13]
```diff
@@ -0,0 +1,634 @@
...
+            self.range_list.as_bytes(),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct RangeOrValueULE(PlainOldULE<8>);
+
+unsafe impl ULE for RangeOrValueULE {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:407]: Nit: Please add a safety comment for `unsafe impl`s like this one

File: components/plurals/src/rules/runtime/ast.rs [hunk 14]
```diff
@@ -0,0 +1,634 @@
...
+            &[encoded],
+            PlainOldULE::<4>::as_byte_slice(&[self.modulo.as_unaligned()]),
+            self.range_list.as_bytes(),
+        ])
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct RangeOrValueULE(PlainOldULE<8>);
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:405]: Suggestion: no need for a newtype here.  Just set `type ULE = PlainOldULE<8>` on `RangeOrValue` (or use `PairULE` as I suggest below)

File: components/plurals/src/rules/runtime/ast.rs [hunk 15]
```diff
@@ -0,0 +1,634 @@
...
+
+impl AsULE for RangeOrValue {
+    type ULE = RangeOrValueULE;
+
+    #[inline]
+    fn as_unaligned(self) -> Self::ULE {
+        let mut result = [0; 8];
+        match self {
+            Self::Range(start, end) => {
+                let start_bytes = start.to_be_bytes();
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:423]: Nit, here and below: it's weird, although not wrong, that you are using `to_be_bytes()` in a ULE impl.  I'd just use `to_le_bytes()`.  You might even get a small performance boost on LE platforms if you do that.

File: components/plurals/src/rules/runtime/ast.rs [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+unsafe impl ULE for RangeOrValueULE {
+    type Error = zerovec::ule::ULEError<core::convert::Infallible>;
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        PlainOldULE::<8>::validate_byte_slice(bytes)
+    }
+}
+
+impl AsULE for RangeOrValue {
+    type ULE = RangeOrValueULE;
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:416]: Suggestion: use `type ULE = PairULE<<u32 as AsULE>::ULE, <u32 as AsULE>::ULE>`

File: components/plurals/src/rules/runtime/ast.rs [hunk 17]
```diff
@@ -0,0 +1,634 @@
...
+                RangeOrValueULE(result.into())
+            }
+        }
+    }
+
+    #[inline]
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        let b = unaligned.0 .0;
+        let start = u32::from_be_bytes([b[0], b[1], b[2], b[3]]);
+        let end = u32::from_be_bytes([b[4], b[5], b[6], b[7]]);
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:442]: Thought: This isn't as nice as it could be.  If you take my suggestion to use `PairULE`, it should clean this up.

File: components/plurals/src/rules/runtime/ast.rs [hunk 18]
```diff
@@ -0,0 +1,634 @@
...
+        );
+
+        let fd = fixed_decimal::decimal::FixedDecimal::from(1);
+        let operands = PluralOperands::from(&fd);
+        assert!(test_rule(&rule, &operands),);
+    }
+
+    #[test]
+    fn complex_rule_test() {
+        let input = "n % 10 = 3..4,9 and n % 100 != 10..19,70..79,90..99 or n = 0";
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:575]: Request: add a roundtrip test (into and out of VarULE) for a complex rule like this one

**Review by sffc — CHANGES_REQUESTED** [review]
Next round of feedback.  I finished runtime/ast.rs which seems to be one of the bigger files.

File: components/plurals/src/rules/runtime/ast.rs (hunk 15 above)
Comment (zbraniecki) [reply components/plurals/src/rules/runtime/ast.rs:423]: I followed what I did in `datetime`. It would be nice to settle on the single model, and IIRC `to_be_bytes` was the preferred one when we discussed it in context of datetime ULE.

File: components/plurals/src/rules/runtime/ast.rs (hunk 16 above)
Comment (zbraniecki) [reply components/plurals/src/rules/runtime/ast.rs:416]: It would be nice to have `Debug` on `PairULE`.

File: components/plurals/src/rules/runtime/ast.rs (hunk 15 above)
Comment (sffc) [reply components/plurals/src/rules/runtime/ast.rs:423]: I see the datetime ULE as a special case because we had a unique constraint, which is that the top bit of the first byte had to be unassigned so we could use it for the discriminant, which only BE guaranteed. But for an entire 4-byte number with no special cases, I think we should generally use LE.

File: components/plurals/src/rules/runtime/ast.rs (hunk 16 above)
Comment (sffc) [reply components/plurals/src/rules/runtime/ast.rs:416]: @Manishearth ^^

File: components/plurals/src/rules/runtime/ast.rs (hunk 6 above)
Comment (Manishearth) [reply components/plurals/src/rules/runtime/ast.rs:358]: Hehehe I wrote this code but yes I should have a closer look, will mention when I've reviewed

File: components/plurals/src/rules/runtime/ast.rs [hunk 19]
```diff
@@ -0,0 +1,599 @@
...
+        // eventually we can use the Pointer::Metadata APIs when they stabilize
+        let fake_slice = core::ptr::slice_from_raw_parts(ptr as *const RangeOrValueULE, len_new);
+        let ret = &*(fake_slice as *const Self);
+        debug_assert_eq!(core::mem::size_of_val(ret), core::mem::size_of_val(bytes));
+        ret
+    }
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        RelationULE::validate_andor_polarity_operand(bytes[0])?;
+        let remaining = &bytes[5..];
```
Comment (Manishearth) [inline components/plurals/src/rules/runtime/ast.rs:373]: We should return an error if the length is not at least 5 so that the first two lines of this function don't panic

File: components/plurals/src/rules/runtime/resolver.rs [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+        let relation = relation.as_relation();
+        if left {
+            if relation.and_or == ast::AndOr::Or {
+                return true;
+            } else {
+                left = test_relation(&relation, operands);
+            }
+        } else if relation.and_or == ast::AndOr::Or {
+            left = test_relation(&relation, operands);
+        }
```
Comment (sffc) [inline components/plurals/src/rules/runtime/resolver.rs:21]: Optional: I prefer shallower if/else and combining branches where possible.

```suggestion
        if left && relation.and_or == ast::AndOr::Or {
            return true;
        }
        if left || relation.and_or == ast::AndOr::Or {
            left = test_relation(&relation, operands);
        }
```

File: components/plurals/src/rules/runtime/resolver.rs [hunk 21]
```diff
@@ -0,0 +1,61 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::operands::PluralOperands;
+use crate::rules::runtime::ast;
+
+pub fn test_rule(rule: &ast::Rule, operands: &PluralOperands) -> bool {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/resolver.rs:8]: Suggestion: I suggested this algorithm for resolving the and/or rules, but it is non-obvious.  I think it would be good to add some prose explaining it.

File: components/plurals/src/rules/runtime/resolver.rs [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+                left = test_relation(&relation, operands);
+            }
+        } else if relation.and_or == ast::AndOr::Or {
+            left = test_relation(&relation, operands);
+        }
+    }
+    left
+}
+
+#[inline]
```
Comment (sffc) [inline components/plurals/src/rules/runtime/resolver.rs:26]: Note: Since this is `#[inline]` it is especially important to help out the compiler a bit and combine the branches as suggested above

File: components/plurals/src/rules/runtime/resolver.rs [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+        false
+    };
+    match relation.polarity {
+        ast::Polarity::Negative => !result,
+        ast::Polarity::Positive => result,
+    }
+}
+
+#[inline]
+fn get_value(relation: &ast::Relation, operands: &PluralOperands) -> Option<u64> {
```
Comment (sffc) [inline components/plurals/src/rules/runtime/resolver.rs:45]: Question: why `u64` when we are storing all the data as `u32` ?  There's a bit of a conflict here between `u32` and `u64` that we should think about.

Comment (sffc) [inline components/plurals/src/rules/runtime/resolver.rs:45]: Nit (optional): I think it might be more clear if you inlined `get_value` into `test_relation` in the source code.

File: components/plurals/tests/fixtures/categories.json [hunk 24]
```diff
@@ -3,12 +3,12 @@
     "langid": "ar",
     "plural_type": "Cardinal",
     "categories": [
-      "few",
```
Comment (sffc) [inline components/plurals/tests/fixtures/categories.json:6]: Note: You should revert the changes in this file.

File: utils/zerovec/src/zerovec/mod.rs [hunk 25]
```diff
@@ -256,6 +256,31 @@ where
...
+    /// let nums: &[PlainOldULE<2>] = &[211_u16.as_unaligned(), 281_u16.as_unaligned(),
+    ///                                 421_u16.as_unaligned(), 461_u16.as_unaligned()];
+    ///
+    /// let zerovec = ZeroVec::<u16>::borrowed_from_slice(nums);
+    ///
+    /// assert!(matches!(zerovec, ZeroVec::Borrowed(_)));
+    /// assert_eq!(bytes, zerovec.as_bytes());
+    /// ```
+    #[inline]
+    pub fn borrowed_from_slice(other: &'a [T::ULE]) -> Self {
```
Comment (sffc) [inline utils/zerovec/src/zerovec/mod.rs:280]: Nit: This looks like old code.  I remember reviewing this PR for @Manishearth and we decided to instead recommend that you just use `ZeroVec::Borrowed` directly; we don't need a function for it.  Please remove this.

**Review by sffc — CHANGES_REQUESTED** [review]
Good!  Nothing big, just a few more small things.

File: components/plurals/src/rules/runtime/resolver.rs (hunk 23 above)
Comment (zbraniecki) [reply components/plurals/src/rules/runtime/resolver.rs:45]: I tried it and didn't like it. It made the code less readable to me and had to use more Option methods to work around the lack of `return None`. I prefer separated with inline.

File: components/plurals/tests/fixtures/categories.json (hunk 24 above)
Comment (zbraniecki) [reply components/plurals/tests/fixtures/categories.json:6]: I don't think I should. This test tests `categories()` not `all()`. And `categories()` are in order because they are used by resolver which benefits from `one` being first. I'll keep it for now. I think at the current state lack of consistency between various orders of iterations on categories is acceptable.

**Comment by zbraniecki** [comment]
I applied the next round of feedback and I think I'm ready to clean up the code, document fix tests and put for full review.

But before I do, I noticed something concerning:

The performance is good - `plurals/pluralrules/overview` goes from `27us` on my machine to `24us` for an 11% win! Yay!

But the memory is much less affected.

`unread_emails` on `main`:
```
[memory] > dhat: Total:     15,573 bytes in 12 blocks
[memory] > dhat: At t-gmax: 15,311 bytes in 5 blocks
```

`unread_emails` on `pr-zv`:
```
[memory] > dhat: Total:     15,078 bytes in 4 blocks
[memory] > dhat: At t-gmax: 15,070 bytes in 3 blocks
```

`elevator_flooors` on `main`:
```
[memory] > dhat: Total:     16,601 bytes in 28 blocks
[memory] > dhat: At t-gmax: 15,732 bytes in 15 blocks
```

`elevator_flooors` on `pr-zv`:
```
[memory] > dhat: Total:     15,077 bytes in 4 blocks
[memory] > dhat: At t-gmax: 15,069 bytes in 3 blocks
```

@sffc @Manishearth - is that expected? Is it VZV cost, or should we investigate further?

**Comment by Manishearth** [comment]
Note that those examples use a static data provider so they're both loading all of the data as part of the binary, which is probably where most of that number comes from.

**Comment by Manishearth** [comment]
It would be worth diffing the amount the memory usage increases between the start of `main()` and the construction of the PluralRules object

File: components/plurals/src/rules/runtime/resolver.rs (hunk 23 above)
Comment (zbraniecki) [reply components/plurals/src/rules/runtime/resolver.rs:43]: We are storing `PluralOperands` as a combination of `usize` and `u64` but we are storing `ast::RangeOrValue` as `u32`, so at some point we have to coalesce to compare, and I'm using `u64` for it.

I'm happy to revisit that, but would prefer not to creep the scope of this PR.

File: components/plurals/src/rules/runtime/ast.rs (hunk 12 above)
Comment (sffc) [reply components/plurals/src/rules/runtime/ast.rs:393]: Reminder:

> Nit: Please add a safety comment for `unsafe impl`s like this one

**Comment by sffc** [comment]
LGTM other than my minor comments above!

**Review by sffc — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]
